### PR TITLE
Add retries around SetPathAsSharedOpaqueOutput

### PR DESCRIPTION
**Pull request by almiliMSFT**
Retries are needed because: (Win|Unix).SetPathAsSharedOpaqueOutput does something like:
  - check if file has write access rights
  - if it doesn't, give it those rights
  - mark the file as shared opaque output
  - ...

There is a race between checking and setting access rights, so it is possible that here we check its rights, we see that it has the correct rights, then someone else (e.g., the cache) revokes those rights, and so we fail to mark the file as shared opaque output.

### CODE REVIEW

File: Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs [hunk 1]
```diff
@@ -180,19 +180,56 @@ private static bool IsSharedOpaqueOutputWithFallback(string expandedPath, bool c
             }
         }
 
+        private const int MaxNumberOfAttemptsForMarkingSharedOpaqueOutputs = 3;
+        private static readonly TimeSpan SleepDurationBetweenMarkingAttempts = TimeSpan.FromMilliseconds(100);
```
Comment (smera) [inline Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs]: Isn't 100ms a little bit too much? I'd have said 10ms to begin with

File: Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs [hunk 2]
```diff
@@ -180,19 +180,56 @@ private static bool IsSharedOpaqueOutputWithFallback(string expandedPath, bool c
...
         /// <summary>
         /// Marks a given path as "shared opaque output"
         /// </summary>
+        /// <remarks>
+        /// Retries are needed because: (Win|Unix).SetPathAsSharedOpaqueOutput does something like
+        ///   - check if file has write access rights
+        ///   - if it doesn't, give it those rights
+        ///   - mark the file as shared opaque output
+        ///   - ...
+        /// There is a race between checking and setting access rights, so it is possible that
```
Comment (smera) [inline Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs]: I'm not denying that this is happening, and retries are always good :). But I'm curious about what's really going on. From my understanding the cache cannot be racing with this (the pip ran, so we actually set the marker before pushing this to the cache, or it is a cache hit and then we change the marker during materialization - if the marker is not there yet). And any race between pips mean a double write anyway.

File: Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs (hunk 1 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs]: Will change to 10s and will add a linear backoff

File: Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs (hunk 2 above)
Comment (almiliMSFT) [reply Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs]: All good questions.  I don't know what exactly happened.  In one build I saw, a file was just produced by a COPYFILE pip, and then marking it subsequently failed with EACCESS.  So it is possible that retrying will not help in this case, but right now that's all we can do.

**Pull request merged**